packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -868,6 +887,104 @@ class ExactAssetImage extends AssetBundleImageProvider {
+  cancel,
+
+  /// Indicates that the caller would like to defer the image and be called
+  /// again after the next [DeferringImageProvider.deferDuration] elapses (or
+  /// at the end of current task queue if the duration is [Duration.zero]).
+  defer,
+
+  /// Indicates that the [DeferringImageProvider] should attempt to resolve the
+  /// image.
+  ///
+  /// After returning this action, the [DeferringImageProvider] will not call
+  /// its [DeferringImageProvider.getNextAction] callback anymore.
+  resolve,
+}
+
+/// An [ImageProvider] that can delay resolving an image using its wrapped
+/// [imageProvider] until an arbitrary condition is met.
+///
+/// For example, the [Image] widget uses this provider to defer loading images
+/// if it is the child of a scrollable that is rapidly scrolling. It cancels
+/// resolving the image if its element is evicted from the tree before scrolling
+/// ends, and otherwise waits for scrolling to slow down before actually
+/// resolving the image. This prevents the framework from over-utilizing
+/// hardware resources for images that will never be shown on screen.
+///
+/// If the image is already in the cache, e.g. from [precacheImage] or because
+/// another provider has already resolved it, the [getNextAction] callback will
+/// never fire and the cached image is reused.
+@immutable
+class DeferringImageProvider<T> extends ImageProvider<T> {
+
+  /// Creates a new [DeferringImageProvider].
+  ///
+  /// All parameters must not be null.
+  const DeferringImageProvider({
+    @required this.imageProvider,
+    @required this.getNextAction,
+  }) : assert(imageProvider != null),
+       assert(getNextAction != null);
+
+  /// The wrapped [ImageProvider] to delegate loading the image and creating
+  /// a key for it.
+  ///
+  /// The image provider's load method may never be called if resolution gets
+  /// cancelled.
+  final ImageProvider<T> imageProvider;
+
+  /// A callback used by [resolve] to determine whether resolution should be
+  /// cancelled, deferred, or executed.
+  ///
+  /// This callback is never called if the [ImageCache] already has the image
+  /// present, such as from [precacheImage] or another provider successfully
+  /// loading it.
+  final DeferringImageProviderAction Function() getNextAction;
+
+  @override
+  void resolveForStream(ImageConfiguration configuration, ImageStream stream) {
+    void deferredResolve(Object key) {
+      if (PaintingBinding.instance.imageCache.containsKey(key)) {
+        imageProvider.resolveForStream(configuration, stream);
+        return;
+      }
+      switch (getNextAction()) {
+        case DeferringImageProviderAction.defer:
+          // Get ourselves to the end of any frame callbacks at the beginning of
+          // the next frame, which should be after the scroll animation has
+          // updated our velocity value.
+          SchedulerBinding.instance.scheduleFrameCallback((_) {
```

Comment:
Wondering: should this be utilizing the `rescheduling` argument? It may get rescheduled if you continue to defer...

Reply:
Undone - we can't do this becausea there's asynchronicity involved here

Reply:
rescheduling is only about maintaining the stack traces, no?

Reply:
Correct, but it's documented as not working if there are async calls between rescheduling. Which there are here.